### PR TITLE
Feature/highlight only mode

### PR DESCRIPTION
**Pull request by MarshallOfSound**
Implements what some call a "light" mode but for readabilities sake I am called it "highlight only" mode because that is what it actually does.

It disables all rules that aren't the highlight color and adds in a couple of extra ones to get the right look and feel.

As an example here is what GPM looks like with this mode active

light1
light2

The original "dark style" mode is unaffected by these changes and the old `enable()` method exists and works the exact same way it used to work

/cc @gmusic-utils/gmusic-theme-js  @gmusic-utils/gmusic-js

### CODE REVIEW

**Comment by chrismou** [comment]
Do you happen to have a GPDMP dev build with this integrated?

**Comment by MarshallOfSound** [comment]
The implementation in GPMDP will have to change a "tiny" bit in order to enable this highlight only mode.  Gimme a sec and I'll swing you a mac build

**Comment by MarshallOfSound** [comment]
@chrismou This is very naive implementation (I need to change the settings window to match the styling still) but the general idea is there.  I'll probably have to change the defaults to be Google's orange.

But here is the mac zip
https://mega.nz/#!mVVERDSS!-nTL6bZqrC51XekkLgHjyqF0ckkdtqGqWav6JuEH8_w

File: dist/gmusic-theme.js [hunk 1]
```diff
@@ -1665,15 +1674,40 @@ function _classCallCheck(instance, Constructor) { if (!(instance instanceof Cons
...
+
+        /**
+         * Enables the custom theme in dark mode (All colors)
+         */
+
+      }, {
+        key: "enableAll",
+        value: function enableAll() {
+          this.disable();
+          document.querySelector('html').classList.add(CONSTANTS.CLASS_NAMESPACE);
```
Comment (chrismou) [inline dist/gmusic-theme.js]: How come the switch from `body` to `html`, out of interest?

Just wondering if this has the potential to be a breaking change, in the unlikely event that anyone's extending it and targetting `.gmusic-theme.material` on the body tag

**Comment by chrismou** [comment]
Aside from the question, all looks good to me on the implementation front. Clicking around and everything looks as expected,

On an unrelated note, you're not actually providing the Google orange (#f16c00) as a colour option. :-P

Also the settings menu header colour is slightly off (Photoshop is telling me it's #f17800) but I figure from what you said you know about that one.

File: README.md [hunk 2]
```diff
@@ -56,8 +56,14 @@ included will not be changed from the defaults.  `enabled` is set to false by de
...
 
+#### `enableAll`
+Enables the current custom theme for ALL custom colors.  Everything becomes themed
+
+### `enableHighlight`
+Enables the current custom theme but ONLY uses the custom highlight color.  Everything else remains standard Google Play Music styles.
+
 #### `enable()`
-Enables the current custom theme
+Is a mirror of `enableAll`, present for backwards compatibility
```
Comment (twolfson) [inline README.md]: I think "alias" is a more common term than "mirror" =/

File: lib/_constants.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
 module.exports = {
   CLASS_NAMESPACE: 'gmusic-theme',
+  CLASS_NAMESPACE_LIGHT: 'gmusic-theme-light',
```
Comment (twolfson) [inline lib/_constants.js]: If we are going to go with light or highlight, we should be consistent. Please name this `CLASS_NAMESPACE_HIGHLIGHT` and `gmusic-theme-highlight`

File: lib/style.css [hunk 4]
```diff
@@ -460,3 +460,26 @@ button.goog-buttonset-default:hover {
...
+body.gmusic-theme-light,
+.gmusic-theme-light .material-detail-view .material-container-details .info .description,
+.gmusic-theme-light .song-table th {
+  color: #212121;
+}
+
+.gmusic-theme-light .song-row .song-indicator {
+  background-image: url('ani_equalizer_black.gif');
+  background-size: 40px 40px;
+}
```
Comment (twolfson) [inline lib/style.css:354]: It's not clear to me what is dark theme or shared in the CSS. Maybe we should break this up into 3 files:
- common.css
- dark.css
- highlight.css

We should be able to concatenate them in JS when necesssary

File: src/gmusic-theme.js [hunk 5]
```diff
@@ -44,20 +52,40 @@ window.GMusicTheme = class GMusicTheme {
...
+    this.enableAll();
+  }
+
+  /**
+   * Enables the custom theme in dark mode (All colors)
+   */
+  enableAll() {
+    this.disable();
+    document.querySelector('html').classList.add(CONSTANTS.CLASS_NAMESPACE);
+    this.enabled = 1;
```
Comment (twolfson) [inline src/gmusic-theme.js]: Instead of using magic numbers, I suggest defining them as constants at the top as well:

``` js
const ENABLED_ALL = 1;
// ...
this.enabled = ENABLED_ALL;
// ...
if (this.enabled === ENABLED_ALL) {

}
```

File: src/gmusic-theme.js [hunk 6]
```diff
@@ -44,20 +52,40 @@ window.GMusicTheme = class GMusicTheme {
...
-    this._drawLogo();
+    this.enableAll();
+  }
+
+  /**
+   * Enables the custom theme in dark mode (All colors)
+   */
+  enableAll() {
+    this.disable();
+    document.querySelector('html').classList.add(CONSTANTS.CLASS_NAMESPACE);
```
Comment (twolfson) [inline src/gmusic-theme.js]: Instead of using `querySelector` for `html`, you can use `document.documentElement`

File: src/gmusic-theme.js [hunk 7]
```diff
@@ -132,13 +160,25 @@ window.GMusicTheme = class GMusicTheme {
...
+        .replace(/<<BACK_SECONDARY_O>>/g, this._rgba(this.BACK_SECONDARY, 0.5))
+        .replace(/<<NOTIMPORTANT>> \!important/g, '');
+    }
+    // DEV: Else remove all rules for anything that isn't the highlight color (foreSecondary)
     return styleString
-      .replace(/<<BACK_PRIMARY>>/g, this.BACK_PRIMARY)
-      .replace(/<<BACK_SECONDARY>>/g, this.BACK_SECONDARY)
-      .replace(/<<BACK_HIGHLIGHT>>/g, this.BACK_HIGHLIGHT)
-      .replace(/<<FORE_PRIMARY>>/g, this.FORE_PRIMARY)
+      .replace(/\n.+<<BACK_PRIMARY>>.*;\n/g, '')
```
Comment (twolfson) [inline src/gmusic-theme.js]: `.+` and `.*` is definitely something to avoid using; it's a very greedy and costly regexp. Instead, use something more targetted like `[^;]+`. That being said, why do we have these new matchers?

Comment (twolfson) [inline src/gmusic-theme.js]: Oh, I see from the comment. Yea, instead of deleting a lot of rules, we should not include them altogether by breaking up the CSS file =/

File: lib/_constants.js (hunk 3 above)
Comment (chrismou) [inline lib/_constants.js]: If the names are an issue, I'd suggest we rename them gmusic-theme-light and gmusic-theme-dark.  It's a bit more futureproof, in case we ever decided to expand the "light" styles out to include a set of custom background colours.  "highlight" pigeonholes it as foreground colours only.

**Comment by twolfson** [comment]
Left a bunch of feedback. As you might guess, I'm opposed to calling it highlight since most theming libraries call themselves dark or light.

Also, I think the creation of `enableAll` and `enableHighlight` is unnecessary. People will likely adjust colors at the same time as when they decide to use a "highlight" or "all" theme.

Therefore, it makes the most sense to make it an option in the same location:

``` js
theme.updateTheme(type: 'highlight');
```

If we want enforcement, then that can be done via asserts:

``` js
// Define our constants
const THEME_TYPE_ALL = 'all';
const THEME_TYPE_HIGHLIGHT = 'highlight';
const SUPPORTED_THEME_TYPES = [THEME_TYPE_ALL, THEME_TYPE_HIGHLIGHT];

// Inside the `updateTheme` method
var assert = require('assert');
updateTheme: function (options) {
  // If we received a theme, verify it's valid
  if (options.type) {
    assert.notEqual(SUPPORTED_THEME_TYPES.indexOf(options.type), -1,
      'Expected `options.type` to be in ' + JSON.stringify(SUPPORTED_THEME_TYPES) + ' but it was not.');
  }

  // Update theme properties
  this.BACK_PRIMARY = colorObject.backPrimary || this.BACK_PRIMARY;
  // ...
  this.TYPE = colorObject.type || this.TYPE;
  this.redrawTheme();
}

// We can go even further and allow people to use these constants
GMusicTheme.TYPES = {
  ALL: THEME_TYPE_ALL,
  HIGHLIGHT: THEME_TYPE_HIGHLIGHT
};
theme.updateTheme(type: GMusicTheme.TYPES.HIGHLIGHT);
```

File: dist/gmusic-theme.js (hunk 1 above)
Comment (MarshallOfSound) [inline dist/gmusic-theme.js]: So that the `light` class can go on the body while still being a subclass to the main theme classs

(If that makes any sense) :+1:

**Comment by MarshallOfSound** [comment]
@twolfson I agree with lots of your comments, but disagree with the idea that we should be calling it "light" and "dark".  I specifically did not name them that because that is not what these methods do.  You could use the `enableAll` method and still end up with what most would consider a "light" theme.

Hence I labelled them after what they _literally_ do as opposed to what they will be primarily used for.  Does that make sense?

**Comment by twolfson** [comment]
Yea, it makes sense but I'm coming from an API design perspective. With a high level module like this, the goal should not be to describe the implementation to end users (e.g. in `gmusic.js`, we don't name `setPlaybackTime` as `moveSlider`).

Instead, we communicate the behavior that the user expects/is looking for. Since the library is a theming library and the canonical language is light/dark, I'm pushing for light/dark so it can stay consistent with the canon.

**Comment by chrismou** [comment]
FWIW I'm not sure that applying API design principles to a library like this is entirely right, as this isn't an API - an API definitely shouldn't have methods that change layout, styles, etc.  TBH, I'm not sure how I'd categorise this - a control module maybe?  It's essentially a UI helper.  It doesn't sit above the application, it's part of it.

IMO, this is why your example works for `gmusic.js` and not this - that library should never be moving a slider, it should be providing the tools to set the playback time, and the app calling it should be moving the slider itself, hence the naming. This library has a method called `updateTheme`, which is appropriate because it does just that.

But TBH, if we're talking about making decisions based on something that _might_ happen in the future, I'm willing to go along with whatever call @MarshallOfSound makes on it, as long as proper thought's been given to how else this theme library could be used. If an app could use this library to create a dark theme by activating the "light theme" functionality and passing it different config values, then it probably shouldn't be called "light".

**Comment by twolfson** [comment]
I'm also supportive of whatever @MarshallOfSound decides on as long as it's discussed thoroughly =)

@chrismou I think you might be getting caught up on semantics. When I refer to API design, I mean the interface that a programmer interacts with. This can be a method call or an HTTP call.

That being said, you make a good point about a "dark highlight" and I'm more open to the naming as a result. It does irk me a little still though, maybe we could name it an "accent" color or similar.

I should note that I would still prefer "highlight" (or whatever we call it) to be set via `updateTheme` since most people consider enabled/disabled to be a binary state and for the previously mentioned grouping of updating a theme at the same time as its colors.

**Comment by twolfson** [comment]
I looked up what the terms Google uses for Material design. It turns out they are primary, secondary, and accent.

https://www.google.com/design/spec/style/color.html#color-color-schemes

https://www.google.com/design/spec/style/color.html#color-text-background-colors

Maybe we should reconsider our naming scheme to be more inline with Google. For example:
- Rename `BACK_PRIMARY` to `PRIMARY_COLOR`
- Rename `BACK_SECONDARY` to `SECONDARY_COLOR`
- Rename `BACK_HIGHLIGHT` to `TERTIARY_COLOR`
- Rename `FORE_PRIMARY` to `PRIMARY_TEXT`
- Rename `FORE_SECONDARY` to`ACCENT_COLOR`

**Comment by twolfson** [comment]
If we do reconsider our naming scheme for colors, let's do it in a separate PR to prevent overwhelming this one for reviewers x_x

**Comment by twolfson** [comment]
Also, if we do reconsider our naming scheme for colors, let's add a visual diff test suite so we can verify we have no visual regressions. I left some notes on this in https://github.com/gmusic-utils/gmusic-theme.js/issues/2

**Comment by chrismou** [comment]
I was just referring to the comparison with gmusic.js and the point about it being high level.  I got the feeling you were focussing more on the functionality that generated the dist/ folder files than the dist files themselves, but if I've misunderstood then I stand corrected :blush: 

Agreed on the naming scheme change. I've had to sift through the styles a couple of times to remind myself which is primary and which is secondary (I kept thinking primary was the accent) so yeah, a name change would be welcome if it removed any ambiguity.

We should probably continue this discussion in an issue, as I suspect there'll be plenty of discussion and this chat will end up being buried when the PR gets through :+1: 

Same with the visual diff suite. The QAs at my work use one for testing ad generation changes, and they swear by it.  Not sure which library they use though.

**Comment by MarshallOfSound** [comment]
@twolfson @chrismou 

I just updated the PR. 
- Highlight specific CSS has been moved to its own CSS file to remove the inefficient regexp's
- Couple of small things from @twolfson comments
- Light has been renamed completely to highlight

There are two other changes I want to do but not in this PR (I don't want breaking changes in this PR)
1. Remove `enable` methods and move to the `updateTheme` method
2. Rename all variables to follow the Google standard that @twolfson pointed out.

For now though, I think this PR should solely focus on implementing this feature not changing our API.

File: .eslintrc [hunk 8]
```diff
@@ -1,3 +1,4 @@
 {
   "extends": "airbnb/base",
+  "parser": "babel-eslint"
```
Comment (twolfson) [inline .eslintrc:3]: It seems that `babel-eslint` has slipped into this PR. Can we relocate it to a preceeding PR?

File: lib/generate-stylesheet.js [hunk 9]
```diff
@@ -7,24 +7,29 @@ rework.comments = require('rework-comments');
...
-try {
-  fs.mkdirSync(__dirname + '/../build');
-} catch (e) {
-  // If the build directory exists absorb the error
-}
-// DEV: Write the CSS minified to the temporary CSS file
-fs.writeFileSync(__dirname + '/../build/rework.css', (reworkCSS));
-console.info('Rework CSS Generated'); // eslint-disable-line
+const parseCSSFile = (fileName, rootNamespace) => {
+  const styles = fs.readFileSync(__dirname + '/' + fileName + '.css', 'utf8');
```
Comment (twolfson) [inline lib/generate-stylesheet.js:11]: `brfs` won't be able to precompute this content. We should probably keep it simple and do separate loads at the head of the file:

``` js
const commonCss = fs.readFileSync(__dirname + '/common.css', 'utf8');
const highlightCss = fs.readFileSync(__dirname + '/highlight.css', 'utf8');
```

Comment (twolfson) [inline lib/generate-stylesheet.js:11]: Ah, never mind. This file is used to generate CSS

File: src/gmusic-theme.js [hunk 10]
```diff
@@ -123,8 +165,8 @@ window.GMusicTheme = class GMusicTheme {
   }
 
   _refreshStyleSheet() {
-    // DEV: Take the current style string and put it in the style element in the DOM
-    this.styleElement.innerHTML = this._substituteColors(BASE_CSS);
+    // DEV: Take the current style string and put it in the style element in the DOM)
+    this.styleElement.innerHTML = this._substituteColors((this.enabled === GMusicTheme.STATES.ENABLED_ALL ? BASE_CSS + BASE_HIGHLIGHT_CSS : BASE_HIGHLIGHT_CSS));
```
Comment (twolfson) [inline src/gmusic-theme.js]: This line has grown pretty long. Maybe the ternary would be best suited as another line of code:

```
var cssContent = this.COMMON_CSS;
if (this.enabled === GMusicTheme.STATES.ENABLED_ALL) {
  cssContent += BASE_HIGHLIGHT_CSS;
}
```

**Comment by twolfson** [comment]
I'm a little bit drunk at the moment. However, I have left some comments which drunk me thinks are valid.

On top of this, I have a few other comments:
- The name choices for `common.css` and `highlight.css` seem backwards
  - `highlight.css` should actually be `common.css`
  - `common.css` should be `all.css` or `full.css`
- We should keep naming consistent between CSS and JS
  - When referring to `common.css` content, it should be `COMMON_CSS` not `BASE_CSS`
- Since we are dynamically generating the CSS, there's no need to perform namespacing
  - We can remove `.gmusic-theme-highlight` as both an addition to the body as well as our CSS selectors. It can all be done via the `refreshStyleSheet` logic.
- This PR is including the addition of the new `enable` methods. Therefore, we should be able to transfer their addition to `updateTheme` instead =/

**Comment by twolfson** [comment]
I should also mention that if you really need it, there are hacks to increase CSS specificity without actually increasing specificity:

http://csswizardry.com/2014/07/hacks-for-dealing-with-specificity/

File: .eslintrc (hunk 8 above)
Comment (MarshallOfSound) [inline .eslintrc:3]: It doesn't make sense for it to be in it's own PR as it literally does nothing other than add ES7 syntax support to eslint.  We don't use ES7 until this commit so the history wouldn't make much sense?

**Comment by MarshallOfSound** [comment]
@twolfson Fair call on the CSS files being backwards :+1: 

However we can't remove `.gmusic-theme-highlight` as it is required for when we are going with a "highlight only" style.  Also removing the old `enable` method and the new ones would be a breaking change

File: .eslintrc (hunk 8 above)
Comment (twolfson) [inline .eslintrc:3]: The purpose of breaking down PRs is typically to make the PRs smaller so there's less things for reviewers to consider when reviewing each PR.

Additionally, it makes sure that backwards compatibility doesn't break for library/build upgrades.

However, I think I meant to comment on `plugins` in `Gruntfile` rather than the `.eslintrc` =/

Comment (MarshallOfSound) [inline .eslintrc:3]: In this case I think is is an exception as the addition of this and the `plugins` line in the gruntfile is too allow the new code to work.  It would make sense from a historical standpoint to add dependencies that don't do anything in one PR and then start using them in a different PR?

I think this might be more of a debate into how to use git...

**Comment by twolfson** [comment]
With respect to "highlight only", can you explain why we need `.gmusic-theme-highlight`? I think the either the concatentation or the CSS cascade should take care any issues =/

With respect to `enable`, I think we are talking past each other. Here's the interface I am thinking about:

``` js
var theme = new Theme(); // Defaults to `theme.type = 'all';` (this can be defined on `prototype` for ease of use)
theme.enable(); // Sets `theme.enabled = true;` and redraws them
// Inside of theme redrawing, our conditionals for CSS concatenation
//   will be based on `theme.type` rather than `theme.enabled === THEME_TYPE_ALL`
theme.updateTheme({type: 'highlight'}); // Sets `theme.type = 'highlight';` and redraws theme
// Once again, the conditional for CSS concat will only use `highlight.css`
//  as it is based on `theme.type` not `theme.enabled`
```

**Comment by twolfson** [comment]
I should also mention that we might want to rename "all" to "full" since it semantically feels more accurate (e.g. we are filling in parts of a stylesheet so the complete state would be "full")

File: .eslintrc (hunk 8 above)
Comment (twolfson) [inline .eslintrc:3]: Yea, I think my comment is mostly pedantic. We should be fine for a small project like this.

**Comment by MarshallOfSound** [comment]
Ahhh, OK I wasn't quite getting what you were saying :+1:

**Comment by MarshallOfSound** [comment]
And @twolfson we need the `.gmusic-theme-highlight` because there is a SINGLE change in styling between full and highlight.  In hightlight mode the top bar (with the search bar) becomes the accent color.

In full mode (dark or whatever you call it) it does not.  This is just because one looks right and one doesn't :+1:

**Comment by MarshallOfSound** [comment]
@twolfson Ta da? :smile:

File: src/gmusic-theme.js [hunk 11]
```diff
@@ -2,11 +2,25 @@ const color = require('color');
...
+  BACK_PRIMARY: '#222326',
+  BACK_SECONDARY: '#121314',
+  BACK_HIGHLIGHT: '#615F59',
+  FORE_PRIMARY: '#FFFFFF',
+  FORE_SECONDARY: '#1ED760',
+};
+
+class GMusicTheme {
+  static TYPES = {
+    FULL: 'FULl',
```
Comment (twolfson) [inline src/gmusic-theme.js]: Typo here, we want `FULL` not`FULl` =P

File: README.md [hunk 12]
```diff
@@ -56,8 +56,14 @@ included will not be changed from the defaults.  `enabled` is set to false by de
 
 ### States
 
+#### `enableAll`
```
Comment (twolfson) [inline README.md]: We haven't yet updated the documentation to reflect the `enable`/`updateTheme` changes

**Comment by twolfson** [comment]
Looks great, almost at the finish line.

I think we might be talking past each other again with respect to the `.gmusic-theme-highlight` selector. I'm saying that the highlight specific CSS (i.e. CSS not used by full) should live in a `highlight.css` file. This will lead to 3 CSS files:
- `common.css` - CSS shared by both full and highlight modes
- `full.css` - CSS only used by full mode
- `highlight.css` - CSS only used by highlight mode

Then, we can keep it simple and only add/remove a single `gmusic-class` on body`. Then, in redraws, we generate CSS like:

``` js
// At the top
const COMMON_CSS = fs.readFileSync(__dirname + '/../build/common.css', 'utf8');
const FULL_CSS = fs.readFileSync(__dirname + '/../build/full.css', 'utf8');
const HIGHLIGHT_CSS = fs.readFileSync(__dirname + '/../build/highlight.css', 'utf8');

// In _refreshStyleSheet
let styles = COMMON_CSS;
if (this.type === GMusicTheme.TYPES.FULL) {
  styles += FULL_CSS;
} else if (this.type === GMusicTheme.TYPES.HIGHLIGHT_ONLY) {
  styles += HIGHLIGHT_CSS;
}
this.styleElement.innerHTML = this._substituteColors(styles);
```

File: src/gmusic-theme.js [hunk 13]
```diff
@@ -74,6 +96,7 @@ window.GMusicTheme = class GMusicTheme {
     this.BACK_HIGHLIGHT = colorObject.backHighlight || this.BACK_HIGHLIGHT;
     this.FORE_PRIMARY = colorObject.forePrimary || this.FORE_PRIMARY;
     this.FORE_SECONDARY = colorObject.foreSecondary || this.FORE_SECONDARY;
+    this.type = colorObject.type || this.type;
```
Comment (twolfson) [inline src/gmusic-theme.js:100]: We should probably add an assertion for `colorObject.type` to be kind to developers who got a typo (or miss the deprecation notice for accent when we introduce it):

``` js
static TYPES = {
  FULL: 'FULL',
  HIGHLIGHT_ONLY: 'HIGHLIGHT_ONLY',
};
static VALID_TYPES = [TYPES.FULL, TYPES.HIGHLIGHT_ONLY];

updateTheme(colorObject) {
  if (colorObject.type !== undefined && GMusicTheme.VALID_TYPES.indexOf(colorObject.type) === -1) {
    throw new Error(`\`updateTheme\` expected \`colorObject.type\` to be in ${JSON.stringify(GMusicTheme.VALID_TYPES)} ` + 
      `but it was "${colorObject.type}"`);
  }
```

Comment (twolfson) [inline src/gmusic-theme.js:100]: Oh, silly me. We could have used the object for an even simpler assert:

```
updateTheme(colorObject) {
  if (colorObject.type !== undefined && GMusicTheme.TYPES[colorObject.type] === undefined) {
    let validTypesStr = JSON.stringify(Object.keys(GMusicTheme.TYPES));
    throw new Error(`\`updateTheme\` expected \`colorObject.type\` to be in ${validTypesStr} ` +
      `but it was "${colorObject.type}"`);
  }
```

**Comment by MarshallOfSound** [comment]
@twolfson Finally got around to getting back to this PR.  I've fixed the things you mentioned and I have also changed the default "highlight / accent" color to be Googles orange as suggested by @chrismou 

I think the CSS file structure now is logical and the naming makes more sense that it did so :+1: 

Is this GTGN? :cactus:

File: README.md (hunk 12 above)
Comment (twolfson) [inline README.md]: The documentation still needs updating =/

Comment (MarshallOfSound) [inline README.md]: Woopsie

File: lib/_constants.js [hunk 14]
```diff
@@ -1,3 +1,4 @@
 module.exports = {
   CLASS_NAMESPACE: 'gmusic-theme',
+  CLASS_NAMESPACE_HIGHLIGHT: 'gmusic-theme-highlight',
```
Comment (twolfson) [inline lib/_constants.js]: If we are using concatenation, then we should no longer need this highlighting class (since there's nothing to override via HTML elements).

File: lib/generate-stylesheet.js [hunk 15]
```diff
@@ -7,24 +7,29 @@ rework.comments = require('rework-comments');
...
+    fs.mkdirSync(__dirname + '/../build');
+  } catch (e) {
+    // If the build directory exists absorb the error
+  }
+  // DEV: Write the CSS minified to the temporary CSS file
+  fs.writeFileSync(__dirname + '/../build/' + fileName + '.css', (reworkCSS));
+  console.info('Rework CSS Generated'); // eslint-disable-line
+};
+
+parseCSSFile('highlight', 'body');
```
Comment (twolfson) [inline lib/generate-stylesheet.js]: We can prob use the same namespace for both files (i.e. both `html` or both `body`)

File: lib/highlight.css [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+.more-songs-container {
+  color: <<FORE_SECONDARY>>;
+}
+
+/*
+* Highlight only mode specific things
+* These are prefixed here with ".gmusic-theme-highlight"
+*/
+
+.gmusic-theme-highlight #material-app-bar {
```
Comment (twolfson) [inline lib/highlight.css]: We should be able to remove these `.gmusic-theme-highlight` selectors. They are already in `highlight.css` after all =P

**Comment by twolfson** [comment]
It looks like we buried a comment that hasn't been fully addressed =/ Going to see if I can find it...

File: lib/highlight.css (hunk 16 above)
Comment (MarshallOfSound) [inline lib/highlight.css]: I think you are missing the need the need for this extra class.  There is highlight only CSS which is only needed when in EXPLICIT highlight mode.  So the highlight CSS that is included when you are in dark mode doesn't need this.  Se we need an extra class representing the fact we are in a highlight only state

**Comment by twolfson** [comment]
Alright, so it looks like we agreed on updating `refreshStyleSheet` but haven't done the changes to it. Here are the relevant comments:

https://github.com/gmusic-utils/gmusic-theme.js/pull/17#discussion_r55918566

https://github.com/gmusic-utils/gmusic-theme.js/pull/17#issuecomment-195804501

To extend on them further, it looks like the majority of the lines in `highlight.css` should be relocated to a `common.css`. That is, these lines:

https://github.com/gmusic-utils/gmusic-theme.js/blob/1e8012b2ee0744022a1d26ed68a5bfbd3cf0347d/lib/highlight.css#L1-L171

Then, the concatenation from https://github.com/gmusic-utils/gmusic-theme.js/pull/17#issuecomment-195804501 should be usable:

```
let styles = COMMON_CSS;
if (this.type === GMusicTheme.TYPES.FULL) {
  styles += FULL_CSS;
} else if (this.type === GMusicTheme.TYPES.HIGHLIGHT_ONLY) {
  styles += HIGHLIGHT_CSS;
}
```

File: lib/generate-stylesheet.js (hunk 15 above)
Comment (MarshallOfSound) [inline lib/generate-stylesheet.js]: There was a reason for this :/ 

I have no idea what it was now, but I know there was a reason :cry:

Comment (twolfson) [inline lib/generate-stylesheet.js]: It used to be because we had both the highlight and full CSS being applied to the page at the same time. As a result, we had to use additional CSS selectors to say what to show. Now that we are dynamically creating the CSS to be a highlight or a full render, it's no longer necessary

File: lib/highlight.css (hunk 16 above)
Comment (twolfson) [inline lib/highlight.css]: I think we talked past each other again, after we move most of the CSS in here to a `common.css` and only show `highlight.css` when someone triggers it, then we should no longer need the classes:

https://github.com/gmusic-utils/gmusic-theme.js/pull/17#issuecomment-199137751

Comment (MarshallOfSound) [inline lib/highlight.css]: Jokes, I completely get what you mean now

Goddamit I'm an idiot

**Comment by MarshallOfSound** [comment]
@twolfson  Finally clicked onto what you were saying this whole time :+1: 

I switched up your `if` statement to a `switch` but yeah.  #Igetitnow :+1:

File: README.md [hunk 17]
```diff
@@ -56,7 +56,7 @@ included will not be changed from the defaults.  `enabled` is set to false by de
 
 ### States
 
-#### `enable()`
+#### `enable`
```
Comment (twolfson) [inline README.md]: I think we accidentally deleted the parentheses here. We should prob keep them to be consistent with the rest of the methods

File: README.md [hunk 18]
```diff
@@ -67,19 +67,23 @@ Disables the current custom theme
...
 Updates the colors used in the custom theme and redraws the theme.
 
-- colorObject - `Object` -  A colors object containing `backPrimary`, `backSecondary`,
+- colorObject - `Object` -  A colors object containing `type`, `backPrimary`, `backSecondary`,
 `backHighlight`, `forePrimary` and `foreSecondary` attributes any attribute not
 included will not be changed.
 
+The `type` attribute can be any value from `window.GMusicTheme.TYPES`.  At the moment
+these are `"FULL"` and `"HIGHLIGHT_ONLY"`.  There are constants for these values that
+can be accessed through the types object.  E.g. `window.GMusicTheme.TYPES.HIGHLIGHT_ONLY`
```
Comment (twolfson) [inline README.md:76]: Not required but it might be nice to provide a UI preview of what "highlight" mode looks like as opposed to "full"

File: lib/common.css [hunk 19]
```diff
@@ -0,0 +1,172 @@
+
```
Comment (twolfson) [inline lib/common.css]: We prob don't need these 2 leading newlines in this file =P

File: lib/highlight.css [hunk 20]
```diff
@@ -0,0 +1,23 @@
+/* Use the custom colors like so
+ * <<COLOR_NAME>>
+ * E.g. backSecondary is "<<BACK_SECONDARY>>"
+ */
```
Comment (twolfson) [inline lib/highlight.css:4]: We prob should put this comment on `lib/common.css` or all of the files

File: src/gmusic-theme.js [hunk 21]
```diff
@@ -123,8 +147,19 @@ window.GMusicTheme = class GMusicTheme {
...
+    // DEV: Take the current style string and put it in the style element in the DOM)
+    let styles = COMMON_CSS;
+    switch (this.type) {
+      case GMusicTheme.TYPES.FULL:
+        styles += FULL_CSS;
+        break;
+      case GMusicTheme.TYPES.HIGHLIGHT_ONLY:
+        styles += HIGHLIGHT_CSS;
+        break;
+      default:
```
Comment (twolfson) [inline src/gmusic-theme.js:159]: We prob don't need a default case since it's a noop currently

**Comment by twolfson** [comment]
Looks good. Left a few pedantic comments. Feel free to ship with or without them :shipit:

**Comment by twolfson** [comment]
Also, I know we have been doing a lot of patch releases but this should prob be a minor release =P

**Comment by MarshallOfSound** [comment]
Fixed up a couple of pedantics.

Left the default case as it is a linting thing :cactus: 

I'll land this in a minor once travis finishes

**Comment by chrismou** [comment]
:+1:
